src/WebJobs.Extensions.DurableTask/LinuxAppServiceLogger.cs

```diff
@@ -0,0 +1,88 @@
+    /// This class is utilized by <c>EventSourceListener</c> to write logs corresponding to
+    /// specific EventSource providers.
+    /// </summary>
+    internal class LinuxAppServiceLogger
+    {
+        private const string ConsolePrefix = "MS_DURABLE_FUNCTION_EVENTS_LOGS";
+        private const string LoggingPath = "/var/log/functionsLogs/durableevents.log";
+
+        // if true, we write to console (linux consumption), else to a file (linux dedicated).
+        private readonly bool writeToConsole;
+
+        /// <summary>
+        /// Create a LinuxAppServiceLogger instance.
+        /// </summary>
+        /// <param name="writeToConsole">If true, write to console (linux consumption) else to a file (dedicated).</param>
+        public LinuxAppServiceLogger(bool writeToConsole)
+        {
+            // If writeToConsole is False, we write to a file
+            this.writeToConsole = writeToConsole;
+        }
+
+        /// <summary>
+        /// Given EventSource message data, we generate a JSON-string that we can log.
+        /// </summary>
+        /// <param name="eventData">An EventSource message, usually generated by an EventListener.</param>
+        /// <returns>A JSON-formatted string representing the input.</returns>
+        private string GenerateJsonStr(EventWrittenEventArgs eventData)
+        {
+            var values = eventData.Payload;
+            var keys = eventData.PayloadNames;
+
+            // We pack them into a JSON
+            JObject json = new JObject();
+            for (int i = 0; i < values.Count; i++)
+            {
+                json.Add(keys[i], (JToken)values[i]);
+            }
+
+            // Generate string-representation of JSON
+            string jsonString = json.ToString(Newtonsoft.Json.Formatting.None);
+            return jsonString;
+        }
+
+        /// <summary>
+        /// Log EventSource message data in Linux AppService.
+        /// </summary>
+        /// <param name="eventData">An EventSource message, usually generated by an EventListener.</param>
+        public void Log(EventWrittenEventArgs eventData)
+        {
+            // Generate JSON string to log based on the EventSource message
+            string jsonString = this.GenerateJsonStr(eventData);
+
+            // We write to console in Linux Consumption
+            if (this.writeToConsole)
+            {
+                // We're ignoring exceptions in the unobserved Task
+                string consoleLine = ConsolePrefix + " " + jsonString;
+                Task unused = Console.Out.WriteLineAsync(consoleLine);
+            }
+            else
+            {
+                // We write to a file in Linux Dedicated
+                using (var writer = new StreamWriter(LoggingPath, append: true))
+                {
+                    // We're ignoring exceptions in the unobserved Task
+                    Task unused = writer.WriteLineAsync(jsonString);
+                }
```

Comment:
So I just now noticed two problems with this block of code:
1. Because we're not awaiting the task, there could be a race condition where we dispose the `StreamWriter` before we finish writing to the file. This could cause us to lose logs due to exceptions.
2. I'm a little worried about the performance impact of opening and closing file handles for every log statement by creating and disposing the `StreamWriter`. Does other Linux dedicated logging do this? I'm wondering if we need to consider keeping the file open indefinitely in append mode to reduce disk I/O. The risk there, however, is that I don't know what impact this might have on the infrastructure that reads the log files while the app is running.

So (1) above should be consider blocking. (2) would be non-blocking but something that makes me a little nervous. We should probably plan on doing some performance testing comparing Windows and Linux to see if there is a noticeable difference before and after this change.

Reply:
Ahh, I think you're right.

@cgillum the Functions Host approach appears to block via `await` but also batches its messages such that it's only writing in 30 second intervals. So it's a compromise. You can find their file logger implementation here: https://github.com/Azure/azure-functions-host/blob/35cf323fa3464a08b410a518bcab006e801301fe/src/WebJobs.Script.WebHost/Diagnostics/LinuxAppServiceFileLogger.cs

I can certainly use this approach, in fact, I had a port of it in my _old_ telemetry PR here: https://github.com/Azure/azure-functions-durable-extension/pull/1391/files


It also reminds me that, as part of the logging setup, I need to create a directory in the linux host when running on linux dedicated, I'll add that change shortly.

but ended up forgetting about it.

Reply:
Please let me know if you'd prefer I go straight away for the linux host approach to file logging

Reply:
I ended up going for the continuation approach, as we discussed. I'm likely to add in a few more changes before merging, but I believe this is mostly done now.